### ISSUE TITLE
v23.09 LV2 Windows Creates Files Everywhere, Or Maybe in the Current Working Directory

### ISSUE THREAD

**Issue by mfisher31** [DISTRHO/Cardinal#587]
This report is for my app(s), but probably applies to any LV2 host on windows.

When loading the LV2 version on Windows with Element running within VSCode,  Cardinal creates meta data all over the place.  See screenshot below of Git changes.  All of the `*\Cardinal\*` files were created by loading the plugin.

![Screenshot 2023-10-04 221916](https://github.com/DISTRHO/Cardinal/assets/880111/255f9b12-42c5-43f0-af40-49ec67ea7d2d)

Other Notes:

- For what it's worth, VST3 seems perfectly fine inside Element.
- I'm using the release version installer, not the *.zip version.
-

**Comment by mfisher31** [comment]
Should note, I'm using the release version installer, not the *.zip version.

**Comment by falkTX** [comment]
seems that finding the user dir fails, and then it uses the current dir?
the stdout/stderr logs printed by cardinal should tell where it is storing user files

**Comment by falkTX** [comment]
What is your hosting passing as the bundle path during LV2 instantiation? I am thinking it is passing an empty string, which then triggers this issue. Or maybe missing the trailing separator at the end as required by the spec?

```
	/**
	   Instantiate the plugin.
...
	   @param bundle_path Path to the LV2 bundle which contains this plugin
	   binary. It MUST include the trailing directory separator (e.g. '/') so
	   that simply appending a filename will yield the path to that file in the
	   bundle.
...
	LV2_Handle (*instantiate)(const struct _LV2_Descriptor * descriptor,
	                          double                         sample_rate,
	                          const char *                   bundle_path,
	                          const LV2_Feature *const *     features);
```

**Comment by falkTX** [comment]
actually found something related, with a potential fix in adfd2966386a44ca9b2160c251619cdf7198222f

**Comment by mfisher31** [comment]
I'm using `lilv_plugin_instantiate` to instantiate the plugin, so i assume that's to LV2 spec.  Any gotchas for building Cardinal on Windows?  I'd like to verify your change.

**Comment by falkTX** [comment]
there are github actions in place which should have automated builds appear very soon, that is the easiest for testing

**Comment by mfisher31** [comment]
Even better! Thanks

**Comment by mfisher31** [comment]
Ok, tested it out and still getting `Cardinal` folder with json files from the current working directly:  windows running host from command line.  Love this plugin though BTW.

**Comment by falkTX** [comment]
thanks for testing, sounds like I need to try a couple of things on a local build until it works as intended.

**Comment by mfisher31** [comment]
Yeah no problem!  I'll load it up in REAPER, or another installed host and see what happens.  I know little to nothing about windows.  Maybe there's differences between running an app from a CMD prompt as opposed to launching from Explorer.

**Comment by mfisher31** [comment]
I should have asked: where is the expected user data dir?  somewhere in `c:/Users/me/AppData` ?

**Comment by falkTX** [comment]
user dir is $mydocs/Cardinal, the adfd2966386a44ca9b2160c251619cdf7198222f tried to fix that by using an alternative method to find "my documents" folder but still fails in the end

**Comment by mfisher31** [comment]
Ok, got it thanks.   
`[0.000 info CardinalCommon.cpp:582 Initializer] Bundle path: C:/Program Files/Common Files/LV2/Cardinal.lv2/`  

Can't blame `lilv` .  I'll check it out a little further.

**Comment by mfisher31** [comment]
Could be this, regarding the 3rd parameter in `SHGetFolderPathW`  use -1 instead of nullptr.
```
Assigning the hToken parameter a value of -1 indicates the Default User. This enables clients of SHGetFolderPath to find folder locations (such as the Desktop folder) for the Default User. The Default User user profile is duplicated when any new user account is created, and includes special folders such as My Documents and Desktop. Any items added to the Default User folder also appear in any new user account.
```

**Comment by falkTX** [comment]
Issue was a silly missing `break`, that fell-through to a direct empty `return {}` :facepalm: 

Fixed in b7ad2d26271662e273e7a83d24dfd514bf8694ed

Please confirm once the builds are available.

**Comment by mfisher31** [comment]
It works!  I tested [da276ee](https://github.com/DISTRHO/Cardinal/commit/da276ee93811c98571663b2e63309eba4eec6915)

As for REAPER, it didn't create any files, anywhere.  In this new build, all seems to be ok!  Thanks
